switchbot/devices/lock.py

```diff
@@ -1 +1,157 @@
+COMMAND_HEADER = "57"
+COMMAND_GET_CK_IV = f"{COMMAND_HEADER}0f2103"
+COMMAND_UNLOCK = f"{COMMAND_HEADER}0f4e01011080"
+COMMAND_LOCK = f"{COMMAND_HEADER}0f4e01011000"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SwitchbotLock(SwitchbotDevice):
+    """Representation of a Switchbot Lock."""
+
+    def __init__(self, device: BLEDevice, key_id: str, encryption_key: str, interface: int = 0, **kwargs: Any) -> None:
+        if len(key_id) == 0:
+            raise ValueError("key_id is missing")
+        elif len(key_id) != 2:
+            raise ValueError("key_id is invalid")
+        if len(encryption_key) == 0:
+            raise ValueError("encryption_key is missing")
+        elif len(encryption_key) != 32:
+            raise ValueError("encryption_key is invalid")
+        self._iv = None
+        self._cipher = None
+        self._key_id = key_id
+        self._encryption_key = bytearray.fromhex(encryption_key)
+        self._update_timer: asyncio.TimerHandle | None = None
+        self._status_before_update: LockStatus | None = None
+        super().__init__(device, None, interface, **kwargs)
+
+    async def lock(self) -> bool:
+        """Send lock command."""
+        return await self._lock_unlock(COMMAND_LOCK, {LockStatus.LOCKED, LockStatus.LOCKING})
+
+    async def unlock(self) -> bool:
+        """Send unlock command."""
+        return await self._lock_unlock(COMMAND_UNLOCK, {LockStatus.UNLOCKED, LockStatus.UNLOCKING})
+
+    async def _lock_unlock(self, command: str, ignore_statuses: set[LockStatus]) -> bool:
+        status = self.get_lock_status()
+        if status is None:
+            await self.update()
+            status = self.get_lock_status()
+        if status in ignore_statuses:
+            return True
+
+        result = await self._send_command(command)
+        if not self._check_command_result(result, 0, {1}):
+            return False
+        self._set_status_update_timer(status)
+        return True
+
+    def _set_status_update_timer(self, status: LockStatus):
+        if self._update_timer is not None:
+            self._update_timer.cancel()
+        self._status_before_update = status
+        self._update_timer = self.loop.call_later(ACTION_UPDATE_DELAY, self._update_status_from_timer)
+
+    def _update_status_from_timer(self):
+        if self._update_timer is not None:
+            self._update_timer.cancel()
+            self._update_timer = None
+        asyncio.create_task(self._update_status())
+
+    async def _update_status(self):
+        await self.update()
+        status = self.get_lock_status()
+        if status == self._status_before_update or status in {LockStatus.LOCKING, LockStatus.UNLOCKING}:
+            self._set_status_update_timer(status)
+        else:
+            self._status_before_update = None
+
+    async def get_basic_info(self) -> dict[str, Any] | None:
+        """Get device basic settings."""
+        return (await self.get_device_data()).data["data"]
```

Comment:
Is there a way you poll the characteristic for status instead since you are actively connected? The advertisement data tends to be stale when connected to the device on the other SwitchBot devices

Reply:
That's what I'm seeing that it takes quite a bit of time to update - sometimes 2 seconds and sometimes more like 10, but unfortunately I haven't yet seen lock status (and other details from advertisement) in responses from characteristic.

Reply:
If you send https://github.com/Danielhiversen/pySwitchbot/blob/a804c7863b253dfe0b09bdca3cd724d21c201492/switchbot/devices/device.py#L36 do you get back state?

Reply:
I get a response back but as far as I can tell it only contains better percentage, firmware version, one more value I haven't been able to correlate yet and some 0s. I've tried executing this in different states of the lock but it doesn't change at all.